### ISSUE TITLE
Support XDG Device portal

### ISSUE THREAD

**Issue by TheEvilSkeleton** [tenacityteam/tenacity-legacy#438]
- [x] I have read the specified guidelines for issues

**Is your feature request related to a problem? Please describe.**

This issue originated from https://github.com/tenacityteam/tenacity/issues/436.

Since the upcoming Flatpak package has access to the host directory, it'd be really good if we can start using portals to further sandbox Tenacity to increase its security without losing any convenience.

**Describe the solution you'd like**

Now that we're making a Flatpak package, we should consider using [XDG Device portals](https://flatpak.github.io/xdg-desktop-portal/portal-docs.html#gdbus-org.freedesktop.portal.Device). When adding support, at least in Flatpak, Tenacity will ask for permissions and then access your devices when granted.

Here's an example with the Decoder application when it tries to access my camera:

![image](https://user-images.githubusercontent.com/50847364/128630165-f59efcd1-ff12-400d-a5b0-be801dac9872.png)

If you deny access, you will be able to grant permissions in Flatseal or your DE's permission manager.

**Comment by Be-ing** [comment]
What devices would this be used for with Tenacity?

**Comment by TheEvilSkeleton** [comment]
External microphones.

**Comment by Be-ing** [comment]
Wouldn't that be handled by Pipewire?

**Comment by TheEvilSkeleton** [comment]
No, they're different. Device portal is used for permissions, e.g. granting access to your microphone, whereas PipeWire just handles audio.

**Comment by n0toose** [comment]
Relevant documentation: https://snapcraft.io/docs/xdg-desktop-portals

**Comment by nbsp** [comment]
How would this work with lightweight or minimal window managers, on non-Flatpak installs? i3 doesn't have a "permission manager", as far as I know.

**Comment by n0toose** [comment]
> How would this work with lightweight or minimal window managers, on non-Flatpak installs? i3 doesn't have a "permission manager", as far as I know.

Normally speaking, you can use the permission manager of another desktop environment. That's what I do when I use i3.

Bigger problem in question: How do we interchange between the pop-up we already have and the fresh XDG stuff?

**Comment by TheEvilSkeleton** [comment]
> Normally speaking, you can use the permission manager of another desktop environment. That's what I do when I use i3.

Yup, can confirm it on Sway + Plasma.

> Bigger problem in question: How do we interchange between the pop-up we already have and the fresh XDG stuff?

What do you mean by pop-up? I do not really understand the question.

**Comment by TheEvilSkeleton** [comment]
I just made a basic Sway + Arch Linux setup without any GTK or Qt libraries installed on the system. At least on Flatpak, it will use GTK permissions out of the box:

![image](https://user-images.githubusercontent.com/50847364/129464830-97a126d1-4ed6-4611-abde-d14381a87f52.png)

If Tenacity uses device portal for microphone, then it should ask to use the microphone (and possibly other devices too).

**Comment by vchernin** [comment]
I definitely think XDG Desktop Portals, implemented either through libportal or some other way, would be useful for Tenacity.

However, I am not sure if they should be used outside of sandboxes like Flatpak: https://github.com/flatpak/libportal/issues/33. 
> I suggest you re-read the xdg-desktop-portal* issues that were open about the behaviour of the portals whem used by non-sandboxed apps, Matthias wasn't keen on opening up that Pandora's box.

So it might be more appropriate to make XDG Desktop portals optional by default for Linux builds (with Flatpak being the one case where they're used).

Edit: there was some discussion about this [on Matrix](https://matrix.to/#/!RfXaBjokqHAbzZrgHz:matrix.org/$OLFI8CqRegFeKFKUaNmovXYNr7BAfg_pm71ruNSfpkk?via=matrix.org) (see above and below discussion).

Edit 2: Actually portals should be acceptable to use for non-sandboxed apps. https://github.com/flatpak/libportal/issues/33#issuecomment-905779376
